### PR TITLE
[TR] PIM-3973 Fix behat scenarios

### PR DESCRIPTION
**Pull request by Demango**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | N
| BC breaks?           | N
| CI currently passes? | N
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | N
| Fixed tickets        | ~
| DB schema updated?   | N
| Migration script?    | ~
| Doc PR               | ~

TODO:
- [x] Fix completeness panel for products with no family #2515 
- [x] Fix item doubling (features/product/date.feature:28) *1
- [x] Fix regression of user name in history panel #2503 
- [ ] Add opening of files
- [ ] Add link from product edit page to product group page
- [x] Add extra info to .field-container to be able to determine the exact field container for an attribute #2505 
- [x] Download Pdf button missing #2516 
- [x] Add attribute widget #2513 
- [ ] Add remove file button
- [x] Fix compound field disabling
- [ ] Removing an attribute fails
- [ ] Fix fake changes being detected
- [ ] Fix completeness showing up as "not yet calculated" when disabled  (features/product/completeness.feature:195) *3
- [ ] Fix associations always displaying as 0 products and 0 groups
- [ ] Fix page title
- [ ] Fix ACL enforcement
- [ ] Fix error message regression (features/product/validation/validate_variant_constraints.feature:8)
- [ ] Fix validation
- [ ] Fix product title not changing when changing locale
- [ ] Add extra info to the selector of compound fields
- [ ] Add sequential edit functionality
- [ ] Fix permissions
- [ ] Fix group button localisation
- [ ] Fix price fields being shown when there are no currencies available *4
- [ ] Fix navigation confirmation when downloading pdf
- [ ] Add new option button
- [ ] Add back to grid button

---

*1:  screenshot from 2015-04-10 14 10 42

*3:
screenshot from 2015-04-16 16 46 13

*4:
screenshot from 2015-04-17 12 52 52

### CODE REVIEW

**Comment by sumbobyboys** [comment]
"Not yet calculated" is not used anymore: completeness is calculated at each completeness load. We should never see this message again :)

File: features/Context/AssertionContext.php [hunk 1]
```diff
@@ -286,43 +290,72 @@ public function iShouldSeeHistory(TableNode $table)
...
+            }
+            if (!$row->hasClass('expanded')) {
+                $row->click();
+            }
+            if (isset($data['author'])) {
+                $author = $row->find('css', 'td.author')->getText();
+                assertEquals(
+                    $data['author'],
+                    $author,
+                    sprintf('Expecting the author of version %s to be %s', $data['version'], $data['author'])
```
Comment (filipsalpe) [inline features/Context/AssertionContext.php]: 'Expecting the author of version %s to be %s, got %s' would be better

File: features/Context/AssertionContext.php [hunk 2]
```diff
@@ -286,43 +290,72 @@ public function iShouldSeeHistory(TableNode $table)
...
+
+            if ($matchingRow->find('css', 'td:nth-of-type(2)')->getText() !== $oldValue && $oldValue) {
+                throw $this->createExpectationException(
+                    sprintf('Wrong old value in row %s, expected %s', $data['property'], $newValue)
+                );
+            }
+
+            if (!preg_match(sprintf('/^%s$/', $newValue), $matchingRow->find('css', 'td:last-of-type')->getText())) {
+                throw $this->createExpectationException(
+                    sprintf('Wrong new value in row %s, expected %s', $data['property'], $newValue)
```
Comment (filipsalpe) [inline features/Context/AssertionContext.php]: Could be nice to see the actual value in the error too

File: features/Context/AssertionContext.php [hunk 3]
```diff
@@ -513,14 +546,14 @@ public function iShouldSeeNotifications(TableNode $table)
...
     public function iShouldSeeThatAttributeIsInheritedFromVariantGroup($attribute)
     {
-        $icons = $this->getCurrentPage()->findFieldIcons($attribute);
-        foreach ($icons as $icon) {
-            if ($icon->hasClass('icon-lock')) {
-                return true;
-            }
-        }
+        $footer = $this->getCurrentPage()->findFieldFooter($attribute);
+        $icon = $footer->find('css', 'div:contains("Updated by variant group")');
```
Comment (filipsalpe) [inline features/Context/AssertionContext.php]: it's not an icon, is it?

File: features/Context/AssertionContext.php [hunk 4]
```diff
@@ -534,14 +567,14 @@ public function iShouldSeeThatAttributeIsInheritedFromVariantGroup($attribute)
...
     public function iShouldSeeThatAttributeIsNotInheritedFromVariantGroup($attribute)
     {
-        $icons = $this->getCurrentPage()->findFieldIcons($attribute);
-        foreach ($icons as $icon) {
-            if (!$icon->hasClass('icon-lock')) {
-                return true;
-            }
-        }
+        $footer = $this->getCurrentPage()->findFieldFooter($attribute);
+        $icon = $footer->find('css', 'span:contains("Updated by variant group")');
```
Comment (filipsalpe) [inline features/Context/AssertionContext.php]: same as above

File: features/Context/NavigationContext.php [hunk 5]
```diff
@@ -208,6 +208,7 @@ public function iAmOnTheEntityEditPage($identifier, $page)
         $getter = sprintf('get%s', $page);
         $entity = $this->getFixturesContext()->$getter($identifier);
         $this->openPage(sprintf('%s edit', $page), array('id' => $entity->getId()));
+        $this->wait();
```
Comment (filipsalpe) [inline features/Context/NavigationContext.php]: not necessary

File: features/Context/Page/Base/Base.php [hunk 6]
```diff
@@ -18,6 +18,7 @@ class Base extends Page
     protected $elements = array(
         'Dialog'         => array('css' => 'div.modal'),
         'Title'          => array('css' => '.navbar-title'),
+        'New title'      => array('css' => '.product-title'),
```
Comment (filipsalpe) [inline features/Context/Page/Base/Base.php]: 'Product title'

File: features/Context/Page/Product/Edit.php [hunk 7]
```diff
@@ -33,15 +33,16 @@ public function __construct($session, $pageFactory, $parameters = array())
             array(
                 'Locales dropdown'        => array('css' => '#locale-switcher'),
                 'Locales selector'        => array('css' => '#pim_product_locales'),
-                'Enable switcher'         => array('css' => '#switch_status'),
+                'Scopes dropdown'         => array('css' => '.scope-switcher'),
+                'Enable switcher'         => array('css' => '.status-switcher'),
```
Comment (filipsalpe) [inline features/Context/Page/Product/Edit.php]: 'Status switcher'

File: features/Context/Page/Product/Edit.php [hunk 8]
```diff
@@ -33,15 +33,16 @@ public function __construct($session, $pageFactory, $parameters = array())
             array(
                 'Locales dropdown'        => array('css' => '#locale-switcher'),
                 'Locales selector'        => array('css' => '#pim_product_locales'),
-                'Enable switcher'         => array('css' => '#switch_status'),
+                'Scopes dropdown'         => array('css' => '.scope-switcher'),
```
Comment (filipsalpe) [inline features/Context/Page/Product/Edit.php]: Channel dropdown

File: features/Context/Page/Product/Edit.php [hunk 9]
```diff
@@ -170,34 +259,312 @@ public function findField($name)
...
+        $label->subLabelContent = $subLabelContent;
+
+        return $label;
+    }
+
+    /**
+     * Guesses the type of field identified by $label and returns it.
+     *
+     * Possible identified fields are :
+     * [multiSelect2, simpleSelect2, datepicker, select, wysiwyg, text, compound]
```
Comment (filipsalpe) [inline features/Context/Page/Product/Edit.php]: a bit outdated?

File: features/Context/Page/Product/Edit.php [hunk 10]
```diff
@@ -170,34 +259,312 @@ public function findField($name)
...
+        $field->setValue($value);
+    }
+
+    /**
+     * Fills a date field element with $value, identified by its $label.
+     *
+     * @param \Behat\Mink\Element\NodeElement $label
+     * @param string                          $value
+     */
+    protected function fillDateField($label, $value)
```
Comment (filipsalpe) [inline features/Context/Page/Product/Edit.php]: redundant method

File: features/Context/Page/Product/Edit.php [hunk 11]
```diff
@@ -139,11 +182,57 @@ public function getFieldsCountFor($group)
...
+            case 'price':
+                $this->fillPriceField($label, $value);
+                break;
+            case 'select':
+                $this->fillSelectField($label, $value);
+                break;
+            case 'text':
+                $this->fillTextField($label, $value);
+                break;
+            case 'text':
```
Comment (filipsalpe) [inline features/Context/Page/Product/Edit.php]: textarea?

File: features/Context/Page/Product/Edit.php [hunk 12]
```diff
@@ -170,34 +259,312 @@ public function findField($name)
...
+        return $field;
+    }
+
+    /**
+     * Extracts and return the label NodeElement, identified by $field content and $element
+     *
+     * @param string  $field
+     * @param Element $element
+     *
+     * @return \Behat\Mink\Element\NodeElement
```
Comment (filipsalpe) [inline features/Context/Page/Product/Edit.php]: return NodeElement - there's a use statement for this class

File: features/Context/WebUser.php [hunk 13]
```diff
@@ -213,21 +213,92 @@ public function iShouldSeeTheCategoryUnderTheCategory($not, $child, $parent)
      */
     public function iVisitTheTab($tab)
     {
+        $this->wait();
```
Comment (filipsalpe) [inline features/Context/WebUser.php]: redundant, no?

File: features/Context/WebUser.php [hunk 14]
```diff
@@ -213,21 +213,92 @@ public function iShouldSeeTheCategoryUnderTheCategory($not, $child, $parent)
...
+    }
+
+    /**
      * @param string $group
      *
      * @Given /^I visit the "([^"]*)" group$/
      */
     public function iVisitTheGroup($group)
     {
+        $this->wait();
```
Comment (filipsalpe) [inline features/Context/WebUser.php]: same

File: features/Context/WebUser.php [hunk 15]
```diff
@@ -213,21 +213,92 @@ public function iShouldSeeTheCategoryUnderTheCategory($not, $child, $parent)
...
     }
 
+    /**
+     * @param string $association
+     *
+     * @Given /^I select the "([^"]*)" association$/
+     */
+    public function iSelectTheAssociation($association)
+    {
+        $this->wait();
```
Comment (filipsalpe) [inline features/Context/WebUser.php]: same

File: features/Context/Page/Product/Edit.php [hunk 16]
```diff
@@ -170,34 +248,272 @@ public function findField($name)
...
+    /**
+     * Find a validation tooltip containing a text
+     *
+     * @param string $text
+     *
+     * @return null|Element
+     */
+    public function findValidationTooltip($text)
+    {
+        return $this->find('css', sprintf('span:contains("%s")', $text));
```
Comment (filipsalpe) [inline features/Context/Page/Product/Edit.php]: span inside .validation-errors

File: features/Context/Page/Product/Edit.php [hunk 17]
```diff
@@ -170,34 +248,278 @@ public function findField($name)
...
+                sprintf(
+                    'Could not find field type for field with label "%s".',
+                    $label->getParent()->getHtml()
+                )
+            );
+        }
+    }
+
+    /**
+     * Find a compund field
```
Comment (jmleroux) [inline features/Context/Page/Product/Edit.php]: compound

File: features/Context/Page/Product/Edit.php [hunk 18]
```diff
@@ -170,34 +248,278 @@ public function findField($name)
...
+        $field->setValue($value);
+    }
+
+    /**
+     * Fills a metric field with $value, identified by its $label.
+     *
+     * @param \Behat\Mink\Element\NodeElement $label
+     * @param string                          $value
+     *
+     * @throws InvalidArgumentException
```
Comment (jmleroux) [inline features/Context/Page/Product/Edit.php]: \InvalidArgumentException

File: features/Context/Page/Product/Edit.php [hunk 19]
```diff
@@ -483,26 +839,27 @@ public function checkCompletenessState($channelCode, $localeCode, $state)
...
+    public function checkCompletenessMissingValues($channelCode, $localeCode, $info)
     {
         $completenessCell = $this
             ->findCompletenessCell($channelCode, $localeCode)
-            ->find('css', 'div.progress-cell');
+            ->find('css', 'div.missing');
 
-        if ($info === "Not yet calculated") {
-            if ($info != $completenessCell->getText()) {
+        if ($info == '') {
```
Comment (jmleroux) [inline features/Context/Page/Product/Edit.php]: `$info === ''`

File: features/Context/Page/Product/Edit.php [hunk 20]
```diff
@@ -689,36 +1055,36 @@ public function copySelectedTranslations()
...
-                break;
-            }
-        }
-        if ($columnIdx === 0) {
-            throw new \InvalidArgumentException(sprintf('Column %s not found', $columnCode));
+        $locale = $completenessTable
+            ->find('css', sprintf('span.locale[data-locale="%s"]', $localeCode));
+
+        if (!$locale) {
+            throw new \Exception(sprintf('Could not find completeness for locale "%s".', $localeCode));
```
Comment (jmleroux) [inline features/Context/Page/Product/Edit.php]: missing PHPDoc for `\Exception`

File: features/Context/Page/VariantGroup/Edit.php [hunk 21]
```diff
@@ -11,10 +12,110 @@
...
+
+    /**
+     * @param string $name
+     *
+     * @return NodeElement
+     */
+    public function findField($name)
+    {
+        $currency = null;
+        if (1 === preg_match('/in ((?:.){1,3})$/', $name)) {
```
Comment (jmleroux) [inline features/Context/Page/VariantGroup/Edit.php]: could ```'/in ((?:.){1,3})$/'``` not just be ```'/in (.{1,3})$/'``` ?

File: features/Context/Page/Product/Edit.php [hunk 22]
```diff
@@ -170,34 +248,278 @@ public function findField($name)
...
+            }
+        }
+
+        if ($element) {
+            $label = $element->find('css', sprintf('label:contains("%s")', $labelContent));
+        } else {
+            $label = $this->find('css', sprintf('label:contains("%s")', $labelContent));
+        }
+
+        if (! $label) {
```
Comment (skeleton) [inline features/Context/Page/Product/Edit.php]: !$label

File: features/Context/Page/Product/Edit.php [hunk 23]
```diff
@@ -170,34 +248,278 @@ public function findField($name)
...
+     * Possible identified fields are :
+     * [date, metric, multiSelect, number, price, select, text, textArea]
+     *
+     * @param $label
+     *
+     * @return string
+     */
+    protected function getFieldType($label)
+    {
+        if (null === $label || false === $label instanceof NodeElement) {
```
Comment (skeleton) [inline features/Context/Page/Product/Edit.php]: !$label instanceof NodeElement

File: features/Context/Page/VariantGroup/Edit.php [hunk 24]
```diff
@@ -11,10 +12,110 @@
...
+            $scope = array_shift($words);
+            $name = implode(' ', $words);
+            // Check that it is really a scoped field, not a field with a two word label
+            if (strtolower($scope) === $scope) {
+                return $this->findScopedField($name, $scope);
+            }
+        }
+        $label = $this->find('css', sprintf('label:contains("%s")', $name));
+        if (!$label) {
+            throw new ElementNotFoundException($this->getSession(), 'form label ', 'value', $name);
```
Comment (jmleroux) [inline features/Context/Page/VariantGroup/Edit.php]: PHPDoc

File: features/Context/Page/VariantGroup/Edit.php [hunk 25]
```diff
@@ -11,10 +12,110 @@
...
+    }
+
+    /**
+     * @param string $locale
+     */
+    public function switchLocale($locale)
+    {
+        $elt = $this->getElement('Locales dropdown')->find('css', '.dropdown-toggle');
+        if (!$elt) {
+            throw new \Exception('Could not find locale switcher.');
```
Comment (jmleroux) [inline features/Context/Page/VariantGroup/Edit.php]: PHPDoc

File: features/Context/Page/Product/Edit.php [hunk 26]
```diff
@@ -170,34 +248,278 @@ public function findField($name)
...
+        return $field;
+    }
+
+    /**
+     * Fills a text field element with $value, identified by its $label.
+     *
+     * @param \Behat\Mink\Element\NodeElement $label
+     * @param string                          $value
+     */
+    protected function fillTextField($label, $value)
```
Comment (skeleton) [inline features/Context/Page/Product/Edit.php]: Type hint $label

File: features/Context/Page/Product/Edit.php [hunk 27]
```diff
@@ -170,34 +248,278 @@ public function findField($name)
...
+
+    /**
+     * Fills a simple select2 field with $value, identified by its $label.
+     *
+     * @param \Behat\Mink\Element\NodeElement $label
+     * @param string                          $value
+     *
+     * @throws InvalidArgumentException
+     */
+    protected function fillSelectField($label, $value)
```
Comment (skeleton) [inline features/Context/Page/Product/Edit.php]: idem

File: features/Context/WebUser.php [hunk 28]
```diff
@@ -974,18 +1055,24 @@ public function iRemoveTheFile($field)
     /**
      * @param string $link
      *
-     * @return Step\Given
-     * @Given /^I open "([^"]*)" in the current window$/
+     * @Given /^I open (the file )"([^"]*)" in the current window$/
      */
-    public function iOpenInTheCurrentWindow($link)
+    public function iOpenInTheCurrentWindow($file, $link)
```
Comment (jmleroux) [inline features/Context/WebUser.php]: won't we always have ```$file = 'the file '``` ?

File: features/Context/Page/Product/Edit.php [hunk 29]
```diff
@@ -170,34 +248,278 @@ public function findField($name)
...
+        );
+    }
+
+    /**
+     * Fills a select2 multi-select field with $values, identified by its $label.
+     *
+     * @param \Behat\Mink\Element\NodeElement $label
+     * @param string                          $value
+     */
+    protected function fillMultiSelectField($label, $values)
```
Comment (skeleton) [inline features/Context/Page/Product/Edit.php]: idem

File: features/Context/Page/Product/Edit.php [hunk 30]
```diff
@@ -170,34 +248,278 @@ public function findField($name)
...
+
+    /**
+     * Fills a compound field with $value, by passing the $label
+     *
+     * @param \Behat\Mink\Element\NodeElement $label
+     * @param string                          $value
+     *
+     * @throws ElementNotFoundException
+     */
+    protected function fillCompoundField($label, $value)
```
Comment (skeleton) [inline features/Context/Page/Product/Edit.php]: idem

File: features/Context/Page/Product/Edit.php (hunk 19 above)
Comment (skeleton) [inline features/Context/Page/Product/Edit.php]: '' === $info

File: features/Context/WebUser.php (hunk 28 above)
Comment (filipsalpe) [inline features/Context/WebUser.php]: we will :)

Comment (filipsalpe) [inline features/Context/WebUser.php]: looks like that's the desired behavior but the variable naming is misleading

File: features/product/validation/validate_text_attributes.feature [hunk 31]
```diff
@@ -27,74 +27,90 @@ Feature: Validate text attributes of a product
...
     When I am on the "bar" product page
     And I change the Email to "bar@foo.com"
     And I save the product
     Then I should see validation tooltip "This value is already set on another product."
-    And I should see validation tooltip "There are errors in this tab!"
-    And the "Attributes" tab should be red
+    And there should be 1 error in the "Other" tab
+    # And the "Attributes" tab should be red
+
+  # @skip
```
Comment (filipsalpe) [inline features/product/validation/validate_text_attributes.feature]: ?

**Pull request merged**